db/migrate/20180601160100_add_deb_version_cmp_function_to_db.rb

```diff
@@ -0,0 +1,127 @@
+
+        CREATE OR REPLACE FUNCTION deb_version_cmp_al(_left text, _right text)
+        RETURNS integer
+        AS $$
+        DECLARE
+          lpair text ARRAY[2];
+          rpair text ARRAY[2];
+        BEGIN
+          lpair := ARRAY['', _left];
+          rpair := ARRAY['', _right];
+
+          LOOP
+            IF lpair[2] = '' AND rpair[2] = '' THEN
+              return 0;
+            END IF;
+
+            lpair := regexp_matches(lpair[2], '(.?)(.*)');
+            rpair := regexp_matches(rpair[2], '(.?)(.*)');
+
+            IF lpair[1] = rpair[1] THEN
+              CONTINUE;
+            END IF;
+            IF lpair[1] = '~' THEN
+              RETURN -1;
+            END IF;
+            IF rpair[1] = '~' THEN
+              RETURN 1;
+            END IF;
+            IF lpair[1] = '' THEN
+              RETURN -1;
+            END IF;
+            IF rpair[1] = '' THEN
+              RETURN 1;
+            END IF;
+            IF lpair[1] SIMILAR TO '[a-zA-Z]' THEN
+              IF rpair[1] SIMILAR TO '[a-zA-Z]' AND ascii(lpair[1]) > ascii(rpair[1]) THEN
+                RETURN 1;
+              END IF;
+              RETURN -1;
+            END IF;
+            IF rpair[1] SIMILAR TO '[a-zA-Z]' THEN
+              RETURN 1;
+            END IF;
+            IF ascii(lpair[1]) < ascii(rpair[1]) THEN
+              RETURN -1;
+            END IF;
+            RETURN 1;
+          END LOOP;
+        END;
+        $$ IMMUTABLE STRICT LANGUAGE plpgsql;
+
+        CREATE OR REPLACE FUNCTION deb_version_cmp(_left text, _right text)
+        RETURNS integer
+        AS $$
+        DECLARE
+          lpair text ARRAY[2];
+          rpair text ARRAY[2];
+          res integer;
+        BEGIN
+          lpair := ARRAY['', _left];
+          rpair := ARRAY['', _right];
+
+          LOOP
+            IF lpair[2] = '' AND rpair[2] = '' THEN
+              return 0;
+            END IF;
+
+            lpair := regexp_matches(lpair[2], '([^\\d]*)(.*)');
+            rpair := regexp_matches(rpair[2], '([^\\d]*)(.*)');
+
+            res := deb_version_cmp_al(lpair[1], rpair[1]);
+            IF res != 0 THEN
+              RETURN res;
+            END IF;
+
+            lpair := regexp_matches(lpair[2], '(\\d*)(.*)');
+            rpair := regexp_matches(rpair[2], '(\\d*)(.*)');
+
+            res := deb_version_cmp_num(lpair[1], rpair[1]);
+            IF res != 0 THEN
+              RETURN res;
+            END IF;
+          END LOOP;
+        END
+        $$ IMMUTABLE STRICT LANGUAGE plpgsql;
+  ".freeze
+
+  def self.up
+    unless connection.adapter_name.downcase.include?('sqlite')
+      execute FUNCTIONS
+    end
+  end
+
+  def self.down
+    unless connection.adapter_name.downcase.include?('sqlite')
+      execute "
+        DROP FUNCTION deb_version_cmp(text, text) CASCADE;
+        DROP FUNCTION deb_version_cmp_al(text, text) CASCADE;
+        DROP FUNCTION deb_version_cmp_num(text, text) CASCADE;
+      "
+    end
+  end
+end
```

Comment:
I'd suggest to pull the SQL string out of the migration and putting it into a dedicated folder e.g. `db/functions` or something like that with proper .sql extension for easier editing. And building some migration mechanism around it, e.g. everytime shasum of the function changes new migration is created. A copy of the SQL statements can be probably made, but the "master" file should not live in migrations folder I think.

Reply:
Has anybody got experience with https://github.com/teoljungberg/fx ?
Because it seems to do exactly what you described.